sound/soc/intel/boards/sof_sdw_rt_amp.c

```diff
@@ -196,26 +196,28 @@ int rt_amp_spk_rtd_init(struct snd_soc_pcm_runtime *rtd)
 
 	rt_amp_map = get_codec_name_and_route(rtd, codec_name);
 
+	if (!strstr(card->components, "spk:")) {
+		ret = snd_soc_add_card_controls(card, rt_amp_controls,
+						ARRAY_SIZE(rt_amp_controls));
+		if (ret) {
+			dev_err(card->dev, "%s controls addition failed: %d\n", codec_name, ret);
+			return ret;
+		}
+
+		ret = snd_soc_dapm_new_controls(&card->dapm, rt_amp_widgets,
+					ARRAY_SIZE(rt_amp_widgets));
+		if (ret) {
+			dev_err(card->dev, "%s widgets addition failed: %d\n", codec_name, ret);
+			return ret;
+		}
+	}
```

Comment:
are these various rt712_spk / rt722_spk / rt_amp exclusive or can a combination of these exist together?

I don't follow how this is working...

Reply:
> are these various rt712_spk / rt722_spk / rt_amp exclusive or can a combination of these exist together?
> 

They can be a combination and exist together. For example, one Speaker PCM includes rt712 + rt1316.

Reply:
but in that case, what would the 'spk' string refer to? rt712 or rt1316?

That doesn't seem very logical to me.

Reply:
> but in that case, what would the 'spk' string refer to? rt712 or rt1316?
> 
> That doesn't seem very logical to me.

In that case, the components will contain "spk:rt712 spk:rt1316" The commits won't add new Speaker widgets or controls, but it didn't prevent machine driver adding "spk:" tags to the components string.

Reply:
can we do this somewhere in a generic code? The control/widget addition.
So, in the files we only add string to the components and then in somewhere we do this:
```c

if (strstr(card->components, "spk:")) {
	ret = snd_soc_add_card_controls(card, sdw_spk_controls,
					ARRAY_SIZE(sdw_spk_controls));
	if (ret) {
		dev_err(card->dev, "failed to add spk controls: %d\n", ret);
		return ret;
	}

	ret = snd_soc_dapm_new_controls(&card->dapm, sdw_spk_widgets,
					ARRAY_SIZE(sdw_spk_widgets));
	if (ret) {
		dev_err(card->dev, "failed to add spk widgets: %d\n", ret);
		return ret;
	}
}
```
and lots of duplicated code will be gone, plus a well placed comment would enhance the situation.

Reply:
> can we do this somewhere in a generic code? The control/widget addition. So, in the files we only add string to the components and then in somewhere we do this:
> 
> ```c
> if (strstr(card->components, "spk:")) {
> 	ret = snd_soc_add_card_controls(card, sdw_spk_controls,
> 					ARRAY_SIZE(sdw_spk_controls));
> 	if (ret) {
> 		dev_err(card->dev, "%s: failed to add spk controls: %d\n", __func__, ret);
> 		return ret;
> 	}
> 
> 	ret = snd_soc_dapm_new_controls(&card->dapm, sdw_spk_widgets,
> 					ARRAY_SIZE(sdw_spk_widgets));
> 	if (ret) {
> 		dev_err(card->dev, "%s: failed to add spk widgets: %d\n", __func__, ret);
> 		return ret;
> 	}
> }
> ```
> 
> and lots of duplicated code will be gone, plus a well placed comment would enhance the situation.

Good suggestion. I will work on it.